packages/flutter_tools/lib/src/build_system/targets/deferred_components.dart

```diff
@@ -0,0 +1,159 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+// @dart = 2.8
+
+import 'package:meta/meta.dart';
+
+import '../../android/deferred_components_setup_validator.dart';
+import '../../base/deferred_component.dart';
+import '../../build_info.dart';
+import '../../project.dart';
+import '../build_system.dart';
+import '../depfile.dart';
+import 'android.dart';
+
+/// An enumeration of the checks and tasks available in [DeferredComponentsSetupValidator].
+///
+/// Each of these enumerations corresponds to a method in [DeferredComponentsSetupValidator]
+/// of the same name. This is used to configure which checks and tasks to run in a
+/// [DeferredComponentsSetupValidatorTarget].
+enum DeferredComponentsSetupValidatorTask {
+  /// Runs [DeferredComponentsSetupValidator.checkAndroidDynamicFeature].
+  checkAndroidDynamicFeature,
+  /// Runs [DeferredComponentsSetupValidator.checkAppAndroidManifestComponentLoadingUnitMapping].
+  checkAppAndroidManifestComponentLoadingUnitMapping,
+  /// Runs [DeferredComponentsSetupValidator.checkAndroidResourcesStrings].
+  checkAndroidResourcesStrings,
+  /// Runs [DeferredComponentsSetupValidator.checkAgainstLoadingUnitGolden].
+  checkAgainstLoadingUnitGolden,
+  /// Runs [DeferredComponentsSetupValidator.writeGolden].
+  writeGolden,
+  /// Runs [DeferredComponentsSetupValidator.clearOutputDir].
+  clearOutputDir,
+}
+
+/// Creates a [DeferredComponentsSetupValidator, executes the configured
+/// [tasks], and displays the validator output to the developer if
+/// changes are recommended.
+class DeferredComponentsSetupValidatorTarget extends Target {
+  /// Create an [AndroidAotDeferredComponentsBundle] implementation for a given [targetPlatform] and [buildMode].
+  DeferredComponentsSetupValidatorTarget({
+    this.tasks,
+    this.dependency,
+    this.title,
+    this.exitOnFail = true,
+    String name = 'deferred_components_setup_validator',
+  }) : _name = name;
+
+  /// The list of tasks that this target will execute.
+  ///
+  /// Each task corresponds to a method in [DeferredComponentsSetupValidator].
+  final List<DeferredComponentsSetupValidatorTask> tasks;
```

Comment:
Why do you need an enum to describe the list of tasks to be done, don't you need to run each task every time?

Reply:
No, Not all tasks are run every time.

checkAndroidDynamicFeature, checkAndroidResourcesStrings, and clearOutputDir are run before the build starts as gradle will fail to build correctly if those checks are not satisfied.

checkAppAndroidManifestComponentLoadingUnitMapping, checkAgainstLoadingUnitGolden, and writeGolden are run after gen_snapshot/assemble is complete, as it depends on the output of gen_snapshot and cannot be run before.

The validator is run twice, each time with a different set of tasks.

Reply:
> checkAndroidDynamicFeature, checkAndroidResourcesStrings, and clearOutputDir are run before the build starts as gradle will fail to build correctly if those checks are not satisfied.

IIRC in the design discussion we asked that the tool not perform multiple validations, instead prefering to run a single task and let the build fail. Why is it necessary to diverge from that design here?

Reply:
I wouldn't characterize the split as multiple validations, it is two parts of a single validation that is split for technical reasons that I don't think are avoidable without deep changes in how dart works. To the developer, this still functions the same in that the build will run and will fail if the validation detects anything.

It is not possible for dart/the tool to know exactly what/how loading units are split based off of the codebase alone. The resulting loading units are only known after gen_snapshot, thereby requiring the half of the validation dependent on the generated loading units to be run after gen_snapshot (but before the app is packaged together by gradle).

There is never any repeated validation and each check/task is to be run as soon as it is technically possible (this impl is coming in a future PR)

Reply:
Its not so much repeating the same validation that I'm concerned about, but about making an already complicated process even more complicated. For example, it is totally possible (and documented today) that users can build their android app through gradle directly. If this feature _requires_ running the tool first in all circumstances, we have to add asterisks everywhere and make our users care about an implementation detail they could previously ignore.

If we don't run these steps first, how does gradle fail exactly?

Reply:
So technically, this validator is actually optional (see https://github.com/flutter/flutter/pull/76192/files#diff-720367cae2c038df890f8656788545a617314a0e5af2196aef9763df6e4f3061R57). I will be writing documentation/guides on how to do setup everything manually.

If the developer were to run gradle without the tool, depending on the exact mistake, it would indicate one of the following:

- An android module(s) does not exist
- String resource cannot be found
- abi mismatch between base module and dynamic feature module
- loading units is not assigned

The developer also looses the golden checking capability that warns them if the loading unit structure was changed.

I don't think adding the validator here prevents developers from building via gradle directly.

Reply:
that was more of an example, and not the only way the complexity manifests. Having checks that depend on aspects of the project structure also make it difficult/impossible to evolve or modify the project structure. 

I was under the impression that the golden checking was run after gen_snapshot as part of the build?

Reply:
The golden checking is also optional. The goal is to prevent surprise changes in the loading units, as sometimes importing a new library that makes use of deferred imports can cause unexpected changes in the generated loading units. The golden allows us to detect and notify the developer changes, whether intentional or not, were detected and also provides a place to reference what loading units contain which dart libs.

Without the golden checking, the build can still proceed as normal. The danger is that unexpected loading units can cause certain libs to be available/unavailable without much warning.

I agree that the checks do tend to cement in the project structure. The more rigid checks are only for high level inclusion of particular modules or flutter-engine-custom metadata that we require. I don't make any attempts at reading the dynamic feature android manifest or build.gradle and simply check for existence. This allows more flexibility in how developers want to organize/write the android sections of their app.

If we were to change the checked-for structure, it may require some light additional config changes, but change should still be possible with the system I'm proposing here as I have tried to not enforce anything that is not strictly necessary for the function of deferred components.

Reply:
Why don't you grab an hour or so on my calendar next week and we can talk though this a bit?